src/context.jl

```diff
@@ -1,105 +1,206 @@
 """
-    Context
-
-Stores common summary information for all Imputor types.
+    AbstractContext
 
-# Fields
-* `num::Int`: number of observations
-* `count::Int`: number of missing values found
-* `limit::Float64`: allowable limit for missing values to impute
-* `missing::Function`: returns a Bool if the value counts as missing.
+An imputation context records summary information about missing data for an imputation algorithm.
 """
-mutable struct Context
-    num::Int
-    count::Int
-    limit::Float64
-    missing::Function
-end
-
-Context(limit::Float64, missing::Function=ismissing) = Context(0, 0, limit, missing)
-
-Base.copy(x::Context) = Context(x.num, x.count, x.limit, x.missing)
+abstract type AbstractContext end
 
 """
-    ismissing(ctx::Context, x) -> Bool
+    ismissing(ctx::AbstractContext, x) -> Bool
 
-Uses `ctx.missing` to determine if x is missing. If x is a data row or an abstract array
-then `ismissing` will return true if `ctx.missing` returns true for any element.
+Uses `ctx.is_missing` to determine if x is missing. If x is a named tuple or an abstract array
+then `ismissing` will return true if `ctx.is_missing` returns true for any element.
 The ctx.count is increased whenever whenever we return true and if `ctx.count / ctx.num`
 exceeds our `ctx.limit` we throw an `ImputeError`
 
 # Arguments
 * `ctx::Context`: the contextual information about missing information.
 * `x`: the value to check (may be an single values, abstract array or row)
 """
-function Base.ismissing(ctx::Context, x)
-    missing = if isa(x, DataFrameRow)
-        any(entry -> ctx.missing(entry[2]), pairs(x))
+function Base.ismissing(ctx::AbstractContext, x)
+    missing = if isa(x, NamedTuple)
+        any(entry -> ctx.is_missing(entry[2]), pairs(x))
     elseif isa(x, AbstractArray)
-        any(ctx.missing, x)
+        any(ctx.is_missing, x)
     else
-        ctx.missing(x)
+        ctx.is_missing(x)
     end
```

Comment:
The point was that any context should be able to take an arbitrary missing checking function. This is less important now that `missing` is standard, but many datasets will still use `NaN` or `-9999`.

Reply:
GitHub has cut of the context if this statement. by not showing enough code

I was actually talking about disptching in the type of `x`
So rather than having 
```
if isa(x, NamedTuple)
...
 elseif isa(x, AbstractArray)
...
else
...
end
```

Reply:
Hmmm, maybe, but that seems like an improvement we can make once the deprecation and API changes are in. Similarly to adding `@inbounds`.

Reply:
https://github.com/invenia/Impute.jl/issues/26